scrapy/downloadermiddlewares/robotstxt.py

```diff
@@ -85,27 +87,24 @@ def _logerror(self, failure, request, spider):
                          extra={'spider': spider})
         return failure
 
-    def _parse_robots(self, response, netloc):
+    def _parse_robots(self, response, netloc, spider):
         self.crawler.stats.inc_value('robotstxt/response_count')
         self.crawler.stats.inc_value(
             'robotstxt/response_status_count/{}'.format(response.status))
-        rp = robotparser.RobotFileParser(response.url)
-        body = ''
-        if hasattr(response, 'text'):
-            body = response.text
-        else:  # last effort try
-            try:
-                body = response.body.decode('utf-8')
-            except UnicodeDecodeError:
-                # If we found garbage, disregard it:,
-                # but keep the lookup cached (in self._parsers)
-                # Running rp.parse() will set rp state from
-                # 'disallow all' to 'allow any'.
-                self.crawler.stats.inc_value('robotstxt/unicode_error_count')
-        # stdlib's robotparser expects native 'str' ;
-        # with unicode input, non-ASCII encoded bytes decoding fails in Python2
-        rp.parse(to_native_str(body).splitlines())
-
+        
+        try:
+            rp = load_object(self._parserimpl).from_crawler(self.crawler, response.body)
```

Comment:
why is `_parserimpl` loaded for each request, why not load it in  `__init__`?

Reply:
to clarify: I think that if user put `ROBOTSTXT_PARSER="..."` which can't be loaded (e.g. because a dependency  is not installed), we should fail early, don't even start the crawl.

Reply:
`_parserimpl` is now loaded only once in `__init__`, and at the same place we are checking if any parser dependencies are unmet, if so we raise an exception.